BondageClub/Scripts/ImageCache.js

```diff
@@ -0,0 +1,217 @@
+"use strict";
+
+/**
+ * What state is the image in.
+ * @enum string
+ */
+const CachedImageState = {
+	UNCACHED: 'uncached',
+	LOADING: 'loading',
+	LOADED: 'loaded',
+	ERRORED: 'errored',
+	FAILED: 'failed',
+};
+
+/**
+ * Cached image data
+ */
+class CachedImage {
+	/**
+	 * Create a new cached image.
+	 *
+	 * @param {ImageCache} cache - The cache holding the image
+	 * @param {string} url - The URL for the image to cache
+	 * @param {function(CachedImage):void} [loadCallback] - A function called when the image loads
+	 */
+	constructor(cache, url, loadCallback = null) {
+		/**
+		 * Only used to inform the cache of loading progress
+		 * @type ImageCacheClass
+		 * @private
+		 */
+		this.cache = cache;
+		/** @type string */
+		this.url = url;
+		/** @type CachedImageState */
+		this.state = CachedImageState.UNCACHED;
+		/** @type number */
+		this.errorCount = 0;
+		/** @type {function(CachedImage):void} */
+		this.loadCallback = loadCallback;
+		/** @type HTMLImageElement */
+		this.element = null;
+		/** @type {{ width: number; height: number; texture: WebGLTexture; }} */
+		this.textureInfo = null;
+	}
+
+	/**
+	 * Load the cached image if it's not already available
+	 */
+	load() {
+		if ((this.state == CachedImageState.LOADING) || (this.state == CachedImageState.LOADED))
+			return;
+
+		// Get a new image element to perform the loading for us
+		this.element = new Image();
+
+		// Set up listeners to keep track of the loading process
+		let image = this;
+		this.element.addEventListener("load", function() {
+			image.errorCount = 0;
+			image.state = CachedImageState.LOADED;
+			image.cache._imageStateDidChange(image);
+		});
+
+		this.element.addEventListener("error", function () {
+			image.errorCount += 1;
+			image.state = CachedImageState.ERRORED;
+			image.cache._imageStateDidChange(image);
+		});
+
+		// Start loading
+		this._refresh();
+	}
+
+	/**
+	 * Manually set the cached image to the given element.
+	 * @param {HTMLImageElement} element - The element to set
+	 */
+	setElement(element) {
+		this.element = element;
+		this.state = CachedImageState.LOADED;
+	}
+
+	/**
+	 * Refresh the given image.
+	 */
+	_refresh() {
+		this.state = CachedImageState.LOADING;
+		this.element.src = this.url;
+	}
+
+	/**
+	 * Is the image an asset image?
+	 * @returns boolean
+	 */
+	isAsset() { return (this.url.indexOf("Assets") >= 0); }
+
+	/**
+	 * Is the image loaded?
+	 */
+	isLoaded() { return this.state == CachedImageState.LOADED; }
+}
+
+/**
+ * The class responsible for loading and caching images
+ */
+class ImageCacheClass {
+	constructor() {
+		/** @type {Map<string, CachedImage>} */
+		this.cache = new Map();
+		this.loadedCount = 0;
+	}
+
+	/**
+	 * Get a cached image from the cache.
+	 *
+	 * @param {string} url - The URL of the image to lookup
+	 * @param {function(CachedImage):void} loadCallback - A callback that will be called when the load completes
+	 * @returns {CachedImage?} The cached image, or null
```

Comment:
This doesn't look like it ever returns `null` (or `undefined` for that matter, which is what the `Map.prototype.get` call would return without the provided key). Either the type notation should be `CachedImage | undefined`, or it should just be `CachedImage` (I suspect the former?)
```suggestion
	 * @returns {CachedImage} The cached image
```

Reply:
Yep, came from a previous iteration where you could ask the cache for an object without loading it (had a `load(url)` method). That didn't seem helpful, so both methods got merged.